### PR TITLE
Reconcile refactoring #489

### PR DESCRIPTION
**Pull request by rigazilla**
This redesigns the Reconcile function as describe in #489

### CODE REVIEW

File: pkg/controller/infinispan/infinispan_controller.go [hunk 1]
```diff
@@ -1280,13 +1283,10 @@ func (r *ReconcileInfinispan) serviceForInfinispan(m *infinispanv1.Infinispan) *
...
 	}
 
-	// Set Infinispan instance as the owner and controller
-	controllerutil.SetControllerReference(m, service, r.scheme)
-
 	return service
 }
 
-func (r *ReconcileInfinispan) serviceForDNSPing(m *infinispanv1.Infinispan) *corev1.Service {
+func (r *ReconcileInfinispan) computeDNSService(m *infinispanv1.Infinispan) *corev1.Service {
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1289]:
```suggestion
func (r *ReconcileInfinispan) computePingService(m *infinispanv1.Infinispan) *corev1.Service {
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 2]
```diff
@@ -163,23 +163,43 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
 		return *recResult, err
-	} else {
-		if infinispan.RemoveCondition("preliminaryChecksFailed") {
-			err1 := r.client.Status().Update(context.TODO(), infinispan)
-			reqLogger.Error(err1, "Could not update error conditions")
-		}
+	}
+	if infinispan.RemoveCondition("preliminaryChecksFailed") {
+		err1 := r.client.Status().Update(context.TODO(), infinispan)
+		reqLogger.Error(err1, "Could not update error conditions")
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:169]: @rigazilla maybe we can introduce upodateInfinispan func with reconcile on error and message parameters as we have about 10 infinispan's updates in the code?

File: pkg/controller/infinispan/infinispan_controller.go [hunk 3]
```diff
@@ -239,56 +257,78 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
+	err = r.reconcileService(infinispan, ser, reqLogger)
+	if err != nil {
+		reqLogger.Error(err, "Error in reconcileService", "Infinispan.Namespace", infinispan.Namespace)
+		return reconcile.Result{}, err
+	}
 
-		// StatefulSet created successfully - return and requeue
-		reqLogger.Info("End of the StetefulSet creation")
-		return reconcile.Result{Requeue: true}, nil
+	serDNS := r.computePingService(infinispan)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:290]:
```suggestion
	pingService := r.computePingService(infinispan)
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 4]
```diff
@@ -239,56 +257,78 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
 
-	} else if err != nil {
-		reqLogger.Error(err, "failed to get StatefulSet")
+	err = r.reconcileService(infinispan, serDNS, reqLogger)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:292]:
```suggestion
	err = r.reconcileService(infinispan, pingService, reqLogger)
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 5]
```diff
@@ -239,56 +257,78 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
+	if err != nil {
+		reqLogger.Error(err, "Error in reconcileService for DNS", "Infinispan.Namespace", infinispan.Namespace)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:294]:
```suggestion
		reqLogger.Error(err, "Error in reconcileService for ping DNS", "Infinispan.Namespace", infinispan.Namespace)
```

File: pkg/controller/infinispan/infinispan_controller.go (hunk 2 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:169]: I'm not sure, I can see the pros of saving some lines of code, but I feel the code is more readable with the explicit api calls.

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:169]: Yes, true, but all coding tech like inheritance, funcs, templates, etc. make code less readable but give us more advanced things and reduce the chance of errors. But we can keep this "as is" as reduced number of code lines are not so large.

P.S. I saw at least two Objects update without error handling. Is it by design or something missing?

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:169]: yeah I also noticed that, I'll do a quick check

File: pkg/controller/infinispan/infinispan_controller.go [hunk 6]
```diff
@@ -239,56 +258,78 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
+					return *result, err
+				}
+			} else {
+				ingress := r.computeIngress(infinispan, reqLogger)
+				result, err := r.reconcileIngress(infinispan, ingress, reqLogger)
+				if result != nil {
+					return *result, err
+				}
+			}
+		default:
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:327]: Looks like we can free to remove default section as now Expose.Type validates before Reconcile loop inside the enums

File: pkg/controller/infinispan/infinispan_controller.go [hunk 7]
```diff
@@ -1346,7 +1263,191 @@ func (r *ReconcileInfinispan) serviceExternal(m *infinispanv1.Infinispan) *corev
...
+			{
+				Hosts: []string{ispn.Spec.Expose.Host},
+			}}
+	}
+	return &ingress
+}
+
+// reconcileIngress creates the external Route if needed
+func (r *ReconcileInfinispan) reconcileIngress(ispn *infinispanv1.Infinispan, ingress *networkingv1beta1.Ingress, logger logr.Logger) (*reconcile.Result, error) {
+	i := &networkingv1beta1.Ingress{}
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1431]:
```suggestion
	ingress := &networkingv1beta1.Ingress{}
```
`i` is not so informative and more suitable for iterators

File: pkg/controller/infinispan/infinispan_controller.go [hunk 8]
```diff
@@ -1346,7 +1263,191 @@ func (r *ReconcileInfinispan) serviceExternal(m *infinispanv1.Infinispan) *corev
...
+	ns := types.NamespacedName{
+		Name:      i.Name,
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1433]:
```suggestion
		Name:      ingress.Name,
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 9]
```diff
@@ -1346,7 +1263,191 @@ func (r *ReconcileInfinispan) serviceExternal(m *infinispanv1.Infinispan) *corev
...
+		Namespace: i.Namespace,
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1434]:
```suggestion
		Namespace: ingress.Namespace,
```
But for me 'i' and 's' are not so informative

File: pkg/controller/infinispan/infinispan_controller.go (hunk 8 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:1433]: changed to `ing`, cause `ingress` is already a func arg.

File: pkg/controller/infinispan/infinispan_controller.go (hunk 9 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:1434]: the scope of these vars is very limited. I don't think they can generate confusion.

File: pkg/controller/infinispan/infinispan_controller.go (hunk 8 above)
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:1433]: Yes, agree about `ing` and `ingress`. But in this case search must be based on `ingress` not `ing`

File: pkg/controller/infinispan/infinispan_controller.go (hunk 9 above)
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:1434]: Yes, sure but it was about search by name/namespace for `ingress` but not `i`

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:1434]: ah yeah sure. thanks!

File: pkg/controller/infinispan/infinispan_controller.go (hunk 2 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:170]: I've added some error handling

**Comment by rigazilla** [comment]
I think this PR is big enough. I would leave further refactoring for future PRs.

File: pkg/controller/infinispan/infinispan_controller.go [hunk 10]
```diff
@@ -134,7 +134,8 @@ type ReconcileInfinispan struct {
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
 func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
-	reqLogger.Info("Reconciling Infinispan. Operator Version: " + version.Version)
+	reqLogger.Info("+++++ Reconciling Infinispan. Operator Version: " + version.Version)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:137]:
```suggestion
	reqLogger.Info(fmt.Sprintf("+++++ Reconciling Infinispan. Operator Version: %s", version.Version))
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 11]
```diff
@@ -163,53 +164,61 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
+	configMap, err := r.computeConfigMap(xsite, infinispan)
+	if err != nil {
+		reqLogger.Error(err, "could not create Infinispan configuration")
+		return reconcile.Result{}, err
+	}
+
+	err = r.reconcileConfigMap(infinispan, configMap, reqLogger)
 	if err != nil {
-		reqLogger.Error(err, "Error in computeXSite", "Infinispan.Namespace", infinispan.Namespace)
+		reqLogger.Error(err, "Error in reconcileConfigMap", "Infinispan.Namespace", infinispan.Namespace)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:205]:
```suggestion
		reqLogger.Error(err, "Error in reconcileConfigMap")
```
Looks like we already have an "Infinispan.Namespace" in the root logger

File: pkg/controller/infinispan/infinispan_controller.go [hunk 12]
```diff
@@ -239,56 +248,76 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
-		// StatefulSet created successfully - return and requeue
-		reqLogger.Info("End of the StetefulSet creation")
-		return reconcile.Result{Requeue: true}, nil
+	pingService := r.computePingService(infinispan)
 
-	} else if err != nil {
-		reqLogger.Error(err, "failed to get StatefulSet")
+	err = r.reconcileService(infinispan, pingService, reqLogger)
+	if err != nil {
+		reqLogger.Error(err, "Error in reconcileService for ping DNS", "Infinispan.Namespace", infinispan.Namespace)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:285]:
```suggestion
		reqLogger.Error(err, "Error in reconcileService for ping DNS")
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 13]
```diff
@@ -1346,7 +1047,428 @@ func (r *ReconcileInfinispan) serviceExternal(m *infinispanv1.Infinispan) *corev
...
+						HTTP: &networkingv1beta1.HTTPIngressRuleValue{
+							Paths: []networkingv1beta1.HTTPIngressPath{
+								{
+									Path: "/",
+									Backend: networkingv1beta1.IngressBackend{
+										ServiceName: ispn.Name,
+										ServicePort: intstr.IntOrString{IntVal: consts.InfinispanPort}}}}},
+					}}},
+		}}
+	if ispn.Spec.Security.EndpointEncryption.CertSecretName != "" {
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1204]:
```suggestion
	if ispn.GetEncryptionSecretName() != "" {
```

File: pkg/controller/infinispan/infinispan_controller.go [hunk 14]
```diff
@@ -1346,7 +1047,428 @@ func (r *ReconcileInfinispan) serviceExternal(m *infinispanv1.Infinispan) *corev
...
+}
+
+// reconcileConfigMap creates the configMap for Infinispan if needed
+func (r *ReconcileInfinispan) reconcileConfigMap(ispn *infinispanv1.Infinispan, configMap *corev1.ConfigMap, logger logr.Logger) error {
+	cm := &corev1.ConfigMap{}
+	ns := types.NamespacedName{
+		Name:      configMap.Name,
+		Namespace: configMap.Namespace,
+	}
+	err := r.client.Get(context.TODO(), ns, cm)
```
Comment (dmvolod) [inline pkg/controller/infinispan/infinispan_controller.go:1071]:
```suggestion
	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: configMap.Namespace, Name: configMap.Name}, cm)
```
@rigazilla maybe we can change all places where `ns` are using for compacting code?

**Comment by dmvolod** [comment]
> I think this PR is big enough. I would leave further refactoring for future PRs.

Excellent PR, @rigazilla , just small comments related to code formatting and logging

File: pkg/controller/infinispan/infinispan_controller.go (hunk 14 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go]: > @rigazilla maybe we can change all places where `ns` are using for compacting code?

sounds good, I've changed it all around.

**Comment by rigazilla** [comment]
merged, @dmvolod thanks for your careful review!

**Pull request merged**